src/System.Private.Xml/src/System/Xml/Core/HtmlRawTextWriterGenerator.ttinclude

```diff
@@ -0,0 +1,985 @@
+        private bool _doNotEscapeUriAttributes;
+
+        protected static TernaryTreeReadOnly elementPropertySearch;
+        protected static TernaryTreeReadOnly attributePropertySearch;
+
+        private const int StackIncrement = 10;
+
+<# if (WriterType == RawTextWriterType.Encoded) { #>
+        public <#= ClassName #>(TextWriter writer, XmlWriterSettings settings) : base(writer, settings)
+        {
+            Init(settings);
+        }
+<# } #>
+
+        public <#= ClassName #>(Stream stream, XmlWriterSettings settings) : base(stream, settings)
+        {
+            Init(settings);
+        }
+
+        internal override void WriteXmlDeclaration(XmlStandalone standalone)
+        {
+            // Ignore xml declaration
+        }
+
+        internal override void WriteXmlDeclaration(string xmldecl)
+        {
+            // Ignore xml declaration
+        }
+
+        /// Html rules allow public ID without system ID and always output "html"
+        public override void WriteDocType(string name, string pubid, string sysid, string subset)
+        {
+            Debug.Assert(name != null && name.Length > 0);
+
+            <#= SetTextContentMark(false) #>
+
+            RawText("<!DOCTYPE ");
+
+            // Bug: Always output "html" or "HTML" in doc-type, even if "name" is something else
+            if (name == "HTML")
+                RawText("HTML");
+            else
+                RawText("html");
+
+            if (pubid != null)
+            {
+                RawText(" PUBLIC \"");
+                RawText(pubid);
+                if (sysid != null)
+                {
+                    RawText("\" \"");
+                    RawText(sysid);
+                }
+                <#= BufferName #>[bufPos++] = (<#= BufferType #>)'"';
+            }
+            else if (sysid != null)
+            {
+                RawText(" SYSTEM \"");
+                RawText(sysid);
+                <#= BufferName #>[bufPos++] = (<#= BufferType #>)'"';
+            }
+            else
+            {
+                <#= BufferName #>[bufPos++] = (<#= BufferType #>)' ';
+            }
+
+            if (subset != null)
+            {
+                <#= BufferName #>[bufPos++] = (<#= BufferType #>)'[';
+                RawText(subset);
+                <#= BufferName #>[bufPos++] = (<#= BufferType #>)']';
+            }
+
+            <#= BufferName #>[bufPos++] = (<#= BufferType #>)'>';
+        }
+
+        // For the HTML element, it should call this method with ns and prefix as String.Empty
+        public override void WriteStartElement(string prefix, string localName, string ns)
+        {
+            Debug.Assert(localName != null && localName.Length != 0 && prefix != null && ns != null);
+
+            elementScope.Push((byte)currentElementProperties);
+
+            if (ns.Length == 0)
+            {
+                Debug.Assert(prefix.Length == 0);
+
+                <#= SetTextContentMark(false) #>
```

Comment:
`                 <#= SetTextContentMark(false) #>`
could be at the begining of the line, to remove extra spaces in empty line in generated code. Also without extra line:

```diff
                Debug.Assert(prefix.Length == 0);
-
-                <#= SetTextContentMark(false) #>
+<#= SetTextContentMark(false) #>
```

Reply:
This will have the trade off of either `SetTextContentMark` will need to take an argument to state how much to indent the result, the code will have inconsistent indentation like EncodeChar does now, or it'll always be unindented. I am ok with any of these.